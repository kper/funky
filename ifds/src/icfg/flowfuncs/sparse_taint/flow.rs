use crate::icfg::state::State;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintNormalFlowFunction;

impl SparseTaintNormalFlowFunction {
    pub fn identity<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        fact: &Fact,
        pc: usize,
        defuse: &mut DefUseChain,
        edges: &mut Vec<Edge>,
    ) -> Result<()> {
        let after = defuse.get_next(ctx, function, &fact.belongs_to_var, pc)?;

        for to in after {
            edges.push(Edge::Normal {
                from: fact.apply(),
                to: to,
                curved: false,
            });
        }

        Ok(())
    }
}

impl SparseNormalFlowFunction for SparseTaintNormalFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        variable: &String,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Fact>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut facts = Vec::new();

        let instructions = &function.instructions;
        let instruction = instructions.get(pc);

        if instruction.is_none() {
            debug!("Instruction is none");
            return Ok(Vec::new());
        }

        let instruction = instruction.unwrap();

        debug!("Next instruction is {:?} for {}", instruction, variable);

        let mut nodes = defuse.demand(ctx, function, variable, pc)?;

        match instruction {
            Instruction::Unop(dest, ..) |
            Instruction::Phi(dest, ..) |
            Instruction::BinOp(dest, ..) | 
            Instruction::Assign(dest, ..) => {
                let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                nodes.extend(x.into_iter().map(|x| x.clone()));
            }
            _ => {}
        }
        let nodes = nodes
            .into_iter()
            .filter_map(|x| {
                if x.pc != x.next_pc {
                    Some(x.apply())
                } else {
                    Some(x)
                }
            })
            .collect::<Vec<_>>();

        facts.extend(nodes.into_iter());

        debug!("Out {:#?}", facts);

        Ok(facts)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::icfg::state::State;
    use crate::ir::ast::Program;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_binop() {
        /*
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0
        */

        let func_name = "main".to_string();
        let function = AstFunction {
            name: func_name.clone(),
            definitions: vec![
                "%0".to_string(),
                "%1".to_string(),
                "%2".to_string(),
                "%3".to_string(),
            ],
            instructions: vec![
                Instruction::Const("%0".to_string(), 1.0),
                Instruction::Const("%1".to_string(), 1.0),
                Instruction::BinOp("%2".to_string(), "%0".to_string(), "%1".to_string()),
                Instruction::BinOp("%3".to_string(), "%0".to_string(), "%1".to_string()),
            ],
            ..Default::default()
        };

        let mut graph = Graph::default();
        let mut state = State::default();

        let mut ctx = Ctx {
            graph: &mut graph,
            state: &mut state,
            prog: &Program {
                functions: vec![function.clone()],
            },
            block_resolver: HashMap::default(),
        };

        let pc = 0;

        // fullfilling precondition of `chain.cache()`
        ctx.state.init_function(&function, pc).unwrap();

        let mut defuse = DefUseChain::default();

        let sparse = SparseTaintNormalFlowFunction;

        let facts = sparse
            .flow(&mut ctx, &function, 2, &"%0".to_string(), &mut defuse)
            .unwrap();

        debug!("facts {:#?}", facts);
        assert_eq!(4, facts.len());

        let cmp_facts: Vec<Fact> = vec![
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%0".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 1,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 3,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
            Fact {
                belongs_to_var: "%2".to_string(),
                var_is_global: false,
                var_is_taut: false,
                var_is_memory: false,
                pc: 4,
                next_pc: 4,
                track: 3,
                function: "main".to_string(),
                memory_offset: None,
            },
        ];
        assert_eq!(cmp_facts, facts);
    }
}
