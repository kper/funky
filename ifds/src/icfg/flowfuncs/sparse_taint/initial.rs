use crate::icfg::tabulation::sparse::defuse::DefUseChain;
use crate::icfg::{flowfuncs::*, tabulation::sparse::Ctx};

pub struct SparseTaintInitialFlowFunction;

impl SparseInitialFlowFunction for SparseTaintInitialFlowFunction {
    fn flow<'a>(
        &self,
        ctx: &mut Ctx<'a>,
        function: &AstFunction,
        pc: usize,
        init_facts: &Vec<Fact>,
        defuse: &mut DefUseChain,
    ) -> Result<Vec<Edge>> {
        debug!("Calling init flow for {} with pc {}", function.name, pc);

        let mut edges = Vec::new();

        // We need the offset, because not every
        // instruction is taintable. For example BLOCK and JUMP
        // have no registers. That's why skip to the next one.
        let offset = 0;
        let instructions = &function.instructions;

        let init_fact = init_facts.get(0).context("Cannot find taut")?.clone();

        debug!("init fact is {:#?}", init_fact);

        loop {
            let pc = pc + offset;
            let instruction = instructions.get(pc).context("Cannot find instr")?;
            debug!("Next instruction is {:?}", instruction);

            match instruction {
                Instruction::Const(dest, _)
                | Instruction::Unknown(dest)
                | Instruction::Assign(dest, _)
                | Instruction::Unop(dest, _)
                | Instruction::BinOp(dest, _, _)
                | Instruction::Phi(dest, _, _)
                | Instruction::Conditional(dest, _)
                | Instruction::Load(dest, _, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Call(_callee, _, dests) => {
                    // When the analysis starts at a Call Instruction,
                    // then the function doesn't get called.
                    // But all destinations get automatically tainted.

                    let before = vec![init_fact.clone()];
                    for b in before.into_iter() {
                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), b.pc)
                            .context("Cannot find taut's fact")?;

                        debug!("Next taut is {:#?}", after_taut);

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }
                    }
                    for dest in dests {
                        let after_var = defuse
                            .demand(ctx, &function, dest, pc)
                            .context("Cannot find var's fact")?
                            .into_iter()
                            .map(|x| x.clone())
                            .collect::<Vec<_>>();

                        // append all left sides to the nodes
                        // %2 = binop %0 %1 -- there %2 is the left side
                        let mut appended = Vec::new();
                        let mut append_lhs = |dest: &String| -> Result<()> {
                            defuse.force_remove_if_outdated(function, dest, pc)?;
                            let x = defuse.demand_inclusive(ctx, function, dest, pc)?;
                            appended.extend(x.into_iter().map(|x| x.clone()));

                            Ok(())
                        };

                        for var in after_var.clone().iter() {
                            append_lhs(&var.belongs_to_var)?;
                        }

                        for var in after_var.into_iter().chain(appended) {
                            let applied = var.apply();

                            // after
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Store(_src, offset, _) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let mem = ctx
                            .state
                            .add_memory_var(function.name.clone(), offset.clone() as usize);

                        let after_var = defuse
                            .demand_inclusive(ctx, &function, &mem.name, pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let mut applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: applied.clone(),
                            });
                        }
                    }
                }
                Instruction::Return(_dests) => {
                    let before = vec![init_fact.clone()];

                    for b in before.into_iter() {
                        // The tautological fact was built by the `pacemaker`
                        // and will not be sparsely propagated.

                        let after_taut = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find taut's fact")?;

                        for taut in after_taut.into_iter() {
                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: taut.clone(),
                            });
                        }

                        let after_var = defuse
                            .demand(ctx, &function, &"taut".to_string(), pc)
                            .context("Cannot find var's fact")?;

                        for var in after_var.into_iter() {
                            let applied = var.apply();

                            assert!(applied.pc <= applied.next_pc);

                            edges.push(Edge::Path {
                                from: init_fact.clone().clone(),
                                to: var.clone(),
                            });
                        }
                    }
                }
                Instruction::Block(_) | Instruction::Jump(_) => {
                    panic!("Block or Jump as first instruction is not supported");
                }
                _ => {}
            }

            break;
        }

        Ok(edges)
    }
}
