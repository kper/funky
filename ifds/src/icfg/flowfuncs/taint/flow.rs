use crate::icfg::flowfuncs::*;
use crate::icfg::state::State;

pub struct TaintNormalFlowFunction;

impl NormalFlowFunction for TaintNormalFlowFunction {
    fn flow(
        &self,
        function: &AstFunction,
        graph: &mut Graph,
        pc: usize,
        variable: &String,
        block_resolver: &BlockResolver,
        state: &mut State,
    ) -> Result<Vec<Edge>> {
        debug!(
            "Calling flow for {} with var {} with pc {}",
            function.name, variable, pc
        );

        let mut edges = Vec::new();

        let instructions = &function.instructions;

        let instruction = instructions.get(pc).context("Cannot find instr")?;
        debug!("Next instruction is {:?} for {}", instruction, variable);

        let is_taut = variable == &"taut".to_string();

        match instruction {
            Instruction::Const(reg, _) if reg == variable => {
                //kill
            }
            Instruction::Const(reg, _) if reg != variable && !is_taut => {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Assign(dest, src) if src == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Assign(dest, src) if dest != variable && src != variable => {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                let before = state 
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Assign(_dest, _src) if _dest == variable => {
                //kill
            }
            Instruction::Unop(dest, src) if src == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Unop(dest, src) if dest != variable && src != variable => {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Unop(_dest, _src) if _dest == variable => {
                //kill
            }
            Instruction::BinOp(dest, src1, _src2) if src1 == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src1)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src1)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src1)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::BinOp(dest, _src1, src2) if src2 == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src2)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src2)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src2)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::BinOp(_dest, _src1, _src2) if _dest == variable => {
                //kill
            }
            Instruction::Kill(reg) if variable == reg => {
                // kill
            }
            Instruction::Jump(block) => {
                let jump_to_pc = block_resolver
                    .get(&(function.name.clone(), block.clone()))
                    .with_context(|| format!("Cannot find block to jump to {}", block))?;

                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    *jump_to_pc,
                    variable,
                )?;

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: true,
                    });
                }
            }
            Instruction::Conditional(_reg, jumps) if jumps.len() == 1 => {
                // edge case for an `if`
                // which continues if the condition is not successful
                for block in jumps.iter() {
                    let jump_to_pc = block_resolver
                        .get(&(function.name.clone(), block.clone()))
                        .with_context(|| format!("Cannot find block to jump to {}", block))?;

                    let after = state.add_statement(
                        function,
                        format!("{:?}", instruction),
                        *jump_to_pc,
                        variable,
                    )?;

                    let before = state
                        .get_facts_at(&function.name, pc)?
                        .filter(|x| &x.belongs_to_var == variable)
                        .cloned();

                    for (b, a) in before.zip(after) {
                        edges.push(Edge::Normal {
                            from: b,
                            to: a,
                            curved: true,
                        });
                    }
                }

                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: true,
                    });
                }
            }
            Instruction::Conditional(_reg, jumps) => {
                for block in jumps.iter() {
                    let jump_to_pc = block_resolver
                        .get(&(function.name.clone(), block.clone()))
                        .with_context(|| format!("Cannot find block to jump to {}", block))?;

                    let after = state.add_statement(
                        function,
                        format!("{:?}", instruction),
                        *jump_to_pc,
                        variable,
                    )?;

                    let before = state
                        .get_facts_at(&function.name, pc)?
                        .filter(|x| &x.belongs_to_var == variable)
                        .cloned();

                    for (b, a) in before.zip(after) {
                        edges.push(Edge::Normal {
                            from: b,
                            to: a,
                            curved: true,
                        });
                    }
                }
            }
            Instruction::Phi(dest, src1, _src2) if src1 == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src1)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src1)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src1)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Phi(dest, _src1, src2) if src2 == variable => {
                let mut after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, dest)?;

                let after2 =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, src2)?;

                after.extend(after2);

                let before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src2)
                    .cloned();

                let copy_before = state
                    .get_facts_at(&function.name, pc)?
                    .filter(|x| &x.belongs_to_var == src2)
                    .cloned();

                for (b, a) in (before.chain(copy_before)).zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Phi(dest, src1, src2)
                if dest != variable && src1 != variable && src2 != variable =>
            {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Phi(_dest, _src1, _src2) => {
                //kill
            }
            Instruction::Table(jumps) => {
                for block in jumps.iter() {
                    let jump_to_pc = block_resolver
                        .get(&(function.name.clone(), block.clone()))
                        .with_context(|| format!("Cannot find block to jump to {}", block))?;

                    let after = state.add_statement(
                        function,
                        format!("{:?}", instruction),
                        *jump_to_pc,
                        variable,
                    )?;

                    let before = state
                        .get_facts_at(&function.name, pc)?
                        .into_iter()
                        .filter(|x| &x.belongs_to_var == variable)
                        .cloned();

                    for (b, a) in before.zip(after) {
                        edges.push(Edge::Normal {
                            from: b,
                            to: a,
                            curved: true,
                        });
                    }
                }
            }
            Instruction::Return(dest) if dest.contains(variable) => {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Store(src, offset, i) if variable == src || variable == i => {
                let mem_var =
                    state.add_memory_var("mem".to_string(), function.name.clone(), *offset);

                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    &mem_var.name,
                )?;

                let after_var = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    &variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == src)
                    .cloned();

                for b in before {
                    for a in after.iter().chain(after_var.iter()) {
                        edges.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                    }
                }
            }
            Instruction::Store(src, offset, i)
                if variable != src
                    && variable != i
                    && variable != &"taut".to_string()
                    && variable != &format!("mem@{}", offset) =>
            {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Store(_src, _offset, _i) => {
                // kill
            }
            Instruction::Load(dest, _offset, _i)
                if variable != dest && variable != _i && !variable.starts_with("mem") =>
            {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
            Instruction::Load(dest, offset, _i) => {
                let after =
                    state.add_statement(function, format!("{:?}", instruction), pc + 1, &dest)?;

                let after_var = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    &variable,
                )?;

                // we cannot know which exact variables, because we only
                // know the `offset`. But, we can eliminate all cases
                // where the offset is higher, so they can't be meant.
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| {
                        (x.memory_offset.is_some() && x.memory_offset <= Some(*offset))
                            || _i == variable
                    });

                for b in before {
                    for a in after.iter().chain(after_var.iter()) {
                        edges.push(Edge::Normal {
                            from: b.clone(),
                            to: a.clone(),
                            curved: false,
                        });
                    }
                }
            }
            _ => {
                let after = state.add_statement(
                    function,
                    format!("{:?}", instruction),
                    pc + 1,
                    variable,
                )?;

                // Identity
                let before = state
                    .get_facts_at(&function.name, pc)?
                    .into_iter()
                    .filter(|x| &x.belongs_to_var == variable)
                    .cloned();

                for (b, a) in before.zip(after) {
                    edges.push(Edge::Normal {
                        from: b,
                        to: a,
                        curved: false,
                    });
                }
            }
        }

        Ok(edges)
    }
}
