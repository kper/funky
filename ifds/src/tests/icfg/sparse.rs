use crate::icfg::tikz2::render_to;
use crate::icfg::{state::State, tabulation::sparse::TabulationSparse};
use crate::solver::Request;
use insta::assert_snapshot;
use log::error;

use crate::grammar::*;

use crate::icfg::flowfuncs::sparse_taint::flow::SparseTaintNormalFlowFunction;
use crate::icfg::flowfuncs::sparse_taint::initial::SparseTaintInitialFlowFunction;
use crate::ir::ast::Function as AstFunction;

macro_rules! ir {
    ($name:expr, $req:expr, $ir:expr) => {{
        let mut convert = TabulationSparse::new(
            SparseTaintInitialFlowFunction,
            SparseTaintNormalFlowFunction,
        );

        let prog = ProgramParser::new().parse(&$ir).unwrap();

        let res = convert.visit(&prog, &$req);

        if let Err(err) = res {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));
            panic!("")
        }

        let (graph, state) = res.unwrap();

        let output = render_to(&graph, &state);

        assert_snapshot!(format!("{}_dot", $name), output);

        convert
    }};
}

#[test]
fn test_ir_const() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_const",
        req,
        "
         define test (result 0) (define %0) {
            %0 = 1
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_double_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_double_const",
        req,
        "
         define test (result 0) (define %0 %1 %2){
            %0 = 1
            %1 = 1
            %2 = %0
            %2 = %1
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_simple_store() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_store",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            STORE FROM %0 OFFSET 0 + %0 ALIGN 2 32
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_simple_load() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };

    ir!(
        "test_ir_simple_load",
        req,
        "
         define test (param %0) (result 0) (define %0 %1) {
            %1 = LOAD OFFSET 0 + %0 ALIGN 0
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_chain_assign() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_chain_assign",
        req,
        "
         define test (result 0) (define %0 %1 %2 %3){
            %0 = 1
            %1 = 1
            %2 = %0
            %3 = %2
         };
    "
    );
}

#[ignore]
#[test]
fn test_ir_unop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_unop",
        req,
        "define test (result 0) (define %0 %1) {
            %0 = 1
            %1 = op %0
            %1 = op %0   
        };"
    );
}

#[test]
fn test_ir_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_binop",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %3 = %1 op %0   
        };"
    );
}

#[test]
fn test_ir_binop_offset() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_binop_offset",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = %0 op %1
            %2 = %1 op %0   
        };"
    );
}

#[ignore]
#[test]
fn test_ir_phi() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_phi",
        req,
        "define test (result 0) (define %0 %1 %2) {
            %0 = 1
            %1 = 1
            %2 = phi %0 %1
        };"
    );
}

#[ignore]
#[test]
fn test_ir_killing_op() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    ir!(
        "test_ir_killing_op",
        req,
        "define test (result 0) (define %0 %1 %2)  {
            %0 = 1
            %1 = 1
            KILL %0
            KILL %1
            %2 = 1
        };"
    );
}

#[test]
fn test_ir_if_else_binop() {
    let req = Request {
        variable: None,
        function: "test".to_string(),
        pc: 0,
    };
    let tabulation = ir!(
        "test_ir_if_else_binop",
        req,
        "define test (result 0) (define %0 %1 %2 %3) {
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = %0
            %2 = 3
            GOTO 3
            BLOCK 2
            %1 = 1
            GOTO 3
            BLOCK 3
            %3 = %1 op %0   
        };"
    );

    let scfg = tabulation
        .get_scfg_graph(&"test".to_string(), &"%1".to_string())
        .unwrap();

    let output = scfg
        .edges
        .iter()
        .map(|x| format!("({}) -> ({})", x.get_from().pc, x.to().pc))
        .collect::<Vec<_>>().join("\n");

    assert_snapshot!(format!("{}_scfg_dot", "test_ir_if_else_binop"), output);
}

#[test]
fn test_ir_if_else() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_if_else",
        req,
        "define main (result 0) (define %0 %1 %2 %3) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 1 ELSE GOTO 2 
            BLOCK 1
            %1 = 2
            %2 = 3
            GOTO 3
            BLOCK 2
            %2 = 4
            GOTO 3
            BLOCK 3
            %3 = %2
        };
        "
    );
}

#[test]
fn test_ir_if() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_if",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            IF %1 THEN GOTO 0
            %1 = 2
            %2 = 3
        };
        "
    );
}

#[test]
fn test_ir_loop() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_loop",
        req,
        "define main (result 0) (define %0 %1) {
            BLOCK 0
            %0 = 1
            %1 = 2
            GOTO 0 
        };
        "
    );
}

#[test]
fn test_ir_table() {
    let req = Request {
        variable: None,
        function: "main".to_string(),
        pc: 1,
    };
    ir!(
        "test_ir_table",
        req,
        "define main (result 0) (define %0 %1 %2) {
            BLOCK 0
            %0 = 1
            %1 = 2
            %2 = 3
            BLOCK 1
            %1 = 2
            %2 = 3
            BLOCK 2
            %2 = 4
            TABLE GOTO 0 1 2 ELSE GOTO 2
        };
        "
    );
}
