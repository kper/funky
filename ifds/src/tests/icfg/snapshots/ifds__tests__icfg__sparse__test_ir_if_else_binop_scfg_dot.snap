---
source: src/tests/icfg/sparse.rs
expression: output

---
(3) -> (11)
